Regression test for doctest reporting

Zac-HD opened the pull request:
This pull replaces #4243, and closes issue #4223. 

Well, mostly - it turns out the underlying issue has been fixed on the master branch in the last four months, so really all I'm adding is the regression tests I wrote and related code cleanup.  CC  @tk0miya - this should be much faster to review than expected!

I have added the reporting issue to the changelog, even though this pull is no longer what fixed it, and my name to the authors list for the tests (hopefully that's enough).

Zac-HD committed:
Fix type hint, minor style improvements

Zac-HD committed:
Add ext.autodoc + ext.doctest test root

Zac-HD commented:
Oh dear - the underlying problem is fixed in my local environment, but *not* in CI.  I think I'll need to reapply the patches, which means checking again to see if I can fix the underlying issue instead of applying patches drawn from #4243.  ETA another day or two - let me know if the schedule is tight and I'll prioritise.

tk0miya commented:
It seems tests are failed. Could you check this please?

```
E           assert 'File "dir/inner.rst", line 1, in default' in ['File "dir/inner.rst", line 2, in default', 'File "dir/inner.rst", line 2, in default', 'File "index.rst", line 2, in default', 'File "index.rst", line 5, in default']
```
https://travis-ci.org/sphinx-doc/sphinx/jobs/328654000#L1801

Note: I think this is not a new feature, so we can merge this after feature freeze.

Zac-HD commented:
Yeah, that's the bug - the doctests are actually in four files not two, and not at those line numbers in any case 😭

Next step, I'll try again to reproduce locally...  see my comment above.

tk0miya commented:
Oh, I'd overlooked your last comment. I can't say X-day at this moment because I don't see status of other issues now. But I think this is not a new feature and it seems not breaking changes.
So please go at your pace.

tk0miya commented on sphinx/environment/__init__.py:
```diff
@@ -358,17 +357,16 @@ def merge_info_from(self, docnames, other, app):
-                return filename[:-len(suffix)]
-        # the file does not have docname
+            if filename.endswith(suffix):
+                return filename.rstrip(suffix)
```
I noticed this is not good. `str.rstrip()` is not alternative of `filename[:-len(suffix]`.
```
>>> "rrrr.rst".rstrip(".rst")
''
```

tk0miya commented on sphinx/environment/__init__.py:
```diff
@@ -358,17 +357,16 @@ def merge_info_from(self, docnames, other, app):
 
     def path2doc(self, filename):
-        # type: (unicode) -> unicode
+        # type: (unicode) -> Optional[unicode]
```
why is `filename` optional?

Zac-HD replied:
That's the return value, not the `filename` argument - and `Optional[unicode]` means that it may return a unicode string *or* `None` (if no suffix matches).

tk0miya replied:
oh, sorry. Please forget my comment...

Zac-HD committed:
Get correct filename for each doctest block

This is complicated because doctests may be included from other files,
eg. dosctrings in a Python module.

Zac-HD committed:
Avoid misreporting line number

Zac-HD commented:
Alright, I'm fairly happy with this now.  There are four commits:

- Add tests that doctests report their locations correctly, with real filenames and line numbers
- Tangential refactoring (I was in the area...)
- A nice, principled fix - our reported filename is now based on the block source, rather than the output document
- Paper over the problems with line numbers, by setting to `None` for doctests or subtracting to cancel an off-by-one-error, and making the test weaker.

I'm not delighted by the last bit, but it's still a strict improvement on the status quo and I don't have time to investigate deeply again before the deadline for 1.7.  Assuming that this pull passes tests and review (cc @tk0miya), I'd therefore suggest merging it and leaving a "proper fix" for follow-up work.

Zac-HD commented:
Ping @tk0miya; I think this is ready to merge now 😄

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -366,6 +367,39 @@ def write(self, build_docnames, updated_docnames, method='update'):
+        try:
+            filename = (
+                path.relpath(node.source, self.env.srcdir)
+                .replace(path.sep, '/')
```
Why do you replace separators by `/`?

tk0miya commented on sphinx/ext/doctest.py:
```diff
@@ -366,6 +367,39 @@ def write(self, build_docnames, updated_docnames, method='update'):
+        except Exception:
+            filename = self.env.doc2path(docname, base=None)
+        if PY2:
+            return filename.encode(fs_encoding)
```
Do you know why is this needed? I think this is not needed, and this should return unicode path string.

tk0miya commented on CHANGES:
```diff
@@ -243,6 +243,7 @@ Bugs fixed
 * #4198: autosummary emits multiple 'autodoc-process-docstring' event. Thanks
   to Joel Nothman.
 * #4081: Warnings and errors colored the same when building
+* #4223: doctest: failing tests reported in wrong file, at wrong line
```
Please move this to 1.7.0b3 section.

tk0miya commented on tests/test_ext_doctest.py:
```diff
@@ -55,3 +56,27 @@ def test_is_allowed_version():
+if PY2:
+    test_reporting_with_autodoc = pytest.mark.xfail(
+        reason='node.source points to document (not filename) under Python 2'
+    )(test_reporting_with_autodoc)
```
Please use decorator style instead.

You can give a condition to `xfail` as an argument.
https://docs.pytest.org/en/2.7.3/skipping.html#mark-a-test-function-as-expected-to-fail

tk0miya reviewed (changes requested):
LGTM with nits.

And could you rebase this to 1.7-release branch?
Then I will merge this into 1.7.0 soon.
